SYNTHESIZER: Allow goto-synthesizer accept all CBMC options

qinheping opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

With this PR, `goto-synthesizer` will accept all of CBMC options so that we don't need to update the backend options in `goto-synthesizer` when they are updated in `CBMC_OPTION`.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

zhassan-aws commented on src/goto-synthesizer/goto_synthesizer_parse_options.cpp:
```diff
@@ -61,8 +66,45 @@ int goto_synthesizer_parse_optionst::doit()
+    if(strcmp(cbmc_args[0], "cbmc") != 0)
+    {
+      throw invalid_command_line_argument_exceptiont(
+        "cbmc-cmd must starts with cbmc, e.g., --cbmc-cmd \"cbmc [args]\"",
```
I don't have much background, but why not make the option `--cbmc-options` instead to remove the need to specify `cbmc` (which seems redundant?).

qinheping replied:
`goto-synthesizer` checks its arguments of  that start with `--` against the `SYNTHEZIER_OPT`, and errors out if the argument is not matched any defined option. It will happen if some CBMC option is not one of the synthesizer options.
So we let the argument start with `cbmc` instead to avoid the check.

tautschnig reviewed (changes requested):
Doing string tokenisation and possibly having to deal with quoting/unquoting feels fragile. Could we please try a rather different approach:

1) Move the implementation of `goto_synthesizer_parse_optionst::goto_synthesizer_parse_optionst(int argc, const char **argv)` to `goto_synthesizer_parse_options.cpp`.
2) Keep `#include <cbmc/cbmc_parse_options.h>` as proposed in the current commit.
3) Use `GOTO_SYNTHESIZER_OPTIONS CBMC_OPTIONS` (no operator required between these) as the first argument to `goto_synthesizer_parse_optionst::goto_synthesizer_parse_optionst`.
4) Now `goto-synthesizer` should just accept all of CBMC's options, and will continue to do so even when those change. We just need to find some good solution for the `--help` output, perhaps just say "Accepts all of cbmc's options plus the following: ..."

qinheping commented:
> Doing string tokenisation and possibly having to deal with quoting/unquoting feels fragile. Could we please try a rather different approach:
> 
>     1. Move the implementation of `goto_synthesizer_parse_optionst::goto_synthesizer_parse_optionst(int argc, const char **argv)` to `goto_synthesizer_parse_options.cpp`.
> 
>     2. Keep `#include <cbmc/cbmc_parse_options.h>` as proposed in the current commit.
> 
>     3. Use `GOTO_SYNTHESIZER_OPTIONS CBMC_OPTIONS` (no operator required between these) as the first argument to `goto_synthesizer_parse_optionst::goto_synthesizer_parse_optionst`.
> 
>     4. Now `goto-synthesizer` should just accept all of CBMC's options, and will continue to do so even when those change. We just need to find some good solution for the `--help` output, perhaps just say "Accepts all of cbmc's options plus the following: ..."

Thank you! I also updated the title and description of this PR.

tautschnig reviewed (approved):
Could you please re-add a test, if that's at all possible (i.e., if there is any new behaviour observable)?

qinheping committed:
Allow goto-synthesizer accept all CBMC options

With this commit, goto-synthesizer will accept all of CBMC options so that we don't need to update the backend options in goto-synthesizer when they are updated in CBMC_OPTION.
